### PR TITLE
Add support for Debian Buster build

### PR DESCRIPTION
**Pull request by sousou1**
Changed Dockerfile and job creation for building with Debian Buster.

The following configuration was confirmed to work until the colcon test.
```
branch: , <br/>
use_connext_static: false, <br/>
use_connext_debs: false, <br/>
use_cyclonedds: false, <br/>
use_fastrtps_static: true, <br/>
use_fastrtps_dynamic: false, <br/>
use_opensplice: false, <br/>
ci_branch: master, <br/>
repos_url: , <br/>
supplemental_repos_url: , <br/>
colcon_branch: , <br/>
use_whitespace: false, <br/>
isolated: true, <br/>
ubuntu_distro: buster, <br/>
colcon_mixin_url: https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml, <br/>
cmake_build_type: None, <br/>
build_args: --event-handlers console_cohesion+ console_package_list+ --cmake-args -DINSTALL_EXAMPLES=OFF -DSECURITY=ON, <br/>
test_args: --event-handlers console_direct+ --executor sequential --retest-until-pass 10, <br/>
compile_with_clang: false, <br/>
coverage: false
```

### CODE REVIEW

File: linux_docker_resources/Dockerfile [hunk 1]
```diff
@@ -12,7 +12,9 @@ ARG COMPILE_WITH_CLANG=false
 ENV DEBIAN_FRONTEND noninteractive
 
 RUN apt-get update && apt-get install --no-install-recommends -y locales
-RUN locale-gen en_US.UTF-8
+RUN if test ${UBUNTU_DISTRO} != buster; then locale-gen en_US.UTF-8; \
+  else sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && \
+  dpkg-reconfigure --frontend=noninteractive locales; fi
```
Comment (dirk-thomas) [inline linux_docker_resources/Dockerfile]: Please use two separate `RUN` lines here: one for Ubuntu, one for Debian.

Same below.

File: linux_docker_resources/Dockerfile [hunk 2]
```diff
@@ -73,9 +77,11 @@ RUN if test ${COMPILE_WITH_CLANG} = true; then apt-get update && apt-get install
 RUN apt-get update && apt-get install --no-install-recommends -y gcovr
 
 # Install the OpenSplice binary from the OSRF repositories.
-RUN apt-get update && apt-get install --no-install-recommends -y libopensplice69=6.9.190705+osrf1-1~$UBUNTU_DISTRO
+#RUN if test ${UBUNTU_DISTRO} = buster; then cd /tmp && curl --silent -OL https://github.com/ADLINK-IST/opensplice/releases/download/OSPL_V6_9_190403OSS_RELEASE/PXXX-VortexOpenSplice-6.9.190403OSS-HDE-x86_64.linux-gcc5.4.0-glibc2.23-installer.tar.gz && tar -xzvf PXXX-VortexOpenSplice-6.9.190403OSS-HDE-x86_64.linux-gcc5.4.0-glibc2.23-installer.tar.gz; fi
+#RUN if test ${UBUNTU_DISTRO} = buster; then sed -i "s/<Id>0<\/Id>/<Id>108<\/Id>/" /tmp/HDE/x86_64.linux/etc/config/ospl.xml; fi
```
Comment (dirk-thomas) [inline linux_docker_resources/Dockerfile]: Please remove these two commented lines.

**Comment by dirk-thomas** [comment]
Should `buster` be added here too? https://github.com/ros2/ci/blob/d62ad19a65edfef36086f4fb14d5b5a84ff12d97/job_templates/snippet/property_parameter-definition_common.xml.em#L47

File: job_templates/snippet/property_parameter-definition_common.xml.em [hunk 3]
```diff
@@ -44,7 +44,7 @@ To use the latest released version, use an empty string.</description>
             <a class="string-array">
               <string>@ubuntu_distro</string>
 @{
-choices = ['bionic', 'xenial']
+choices = ['bionic', 'xenial', 'buster']
```
Comment (sousou1) [inline job_templates/snippet/property_parameter-definition_common.xml.em]: I added this line for manual CI.
Should I delete it?

File: linux_docker_resources/Dockerfile (hunk 1 above)
Comment (sousou1) [reply linux_docker_resources/Dockerfile]: Thank you for the review.
Changed as you pointed out.

File: job_templates/snippet/property_parameter-definition_common.xml.em (hunk 3 above)
Comment (nuclearsandwich) [reply job_templates/snippet/property_parameter-definition_common.xml.em]: I think we want to be able to run buster CI if it's available. :+1: for keeping it.

File: linux_docker_resources/Dockerfile [hunk 4]
```diff
@@ -12,7 +12,8 @@ ARG COMPILE_WITH_CLANG=false
 ENV DEBIAN_FRONTEND noninteractive
 
 RUN apt-get update && apt-get install --no-install-recommends -y locales
-RUN locale-gen en_US.UTF-8
+RUN if test ${UBUNTU_DISTRO} != buster; then locale-gen en_US.UTF-8; fi
+RUN if test ${UBUNTU_DISTRO} = buster; then sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && dpkg-reconfigure --frontend=noninteractive locales; fi
```
Comment (nuclearsandwich) [inline linux_docker_resources/Dockerfile]: It's not great for readability, but I think these conditionals should be combined into one RUN invocation unless there's a reason they couldn't be.

File: linux_docker_resources/Dockerfile [hunk 5]
```diff
@@ -29,7 +30,8 @@ RUN echo "deb http://repositories.ros.org/ubuntu/testing/ `lsb_release -cs` main
 RUN echo "Bust Cache for key update 2019-06-08" && curl --silent http://repositories.ros.org/repos.key | apt-key add -
 
 # Add the OSRF repositories to the apt sources list.
-RUN echo "deb http://packages.osrfoundation.org/gazebo/ubuntu `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-latest.list
+RUN if test ${UBUNTU_DISTRO} != buster; then echo "deb http://packages.osrfoundation.org/gazebo/ubuntu `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-latest.list; fi
+RUN if test ${UBUNTU_DISTRO} = buster; then echo "deb http://packages.osrfoundation.org/gazebo/debian-stable `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-latest.list ; fi
```
Comment (nuclearsandwich) [inline linux_docker_resources/Dockerfile]: Same comment as https://github.com/ros2/ci/pull/355/files#diff-a7a7037d3453364950fd80f5f91ed7eeR16

**Comment by nuclearsandwich** [comment]
@sousou1 do you have a Jenkins instance with builds running this Debian support or should we set up a test deployment from this branch to iterate further?

**Comment by sousou1** [comment]
@nuclearsandwich 

Thank you for the review!

I have an instance for jenkins testing. Test again after committing the review!

Should I share the output of the test results?

**Comment by nuclearsandwich** [comment]
> Should I share the output of the test results?

An example build or two wouldn't hurt to share but I don't know that I would have time to look at it. I was asking to check whether the changes you've made have been run as part of a live build yet or if we should get these test-deployed to ci.ros2.org so you could iterate.

**Comment by sousou1** [comment]
@nuclearsandwich 
Thank you for your reply.

> I was asking to check whether the changes you've made have been run as part of a live build yet or if we should get these test-deployed to ci.ros2.org so you could iterate.

Does this mean I have confirmed that my PR works correctly?
In my environment, both bionic and debian Dockerfiles worked correctly.

**Comment by clalancette** [comment]
Given the age of this one, I'm going to close it out.  I don't think we should add Buster support without first adding it to REP-2000.

**Pull request closed without merging**